### PR TITLE
[PLATFORM-1591]: Healthchecks

### PR DESCRIPTION
**Pull request by MaeIsBad**
https://prima-assicurazioni-spa.myjetbrains.com/youtrack/issue/PLATFORM-1591

Fixes https://github.com/primait/localauth0/issues/79

Adds a `localauth0 healthcheck` command and a alpine based image

### CODE REVIEW

File: .github/workflows/cd.yml
Comment (cpiemontese) [inline .github/workflows/cd.yml:1]: The `Create container` script is becoming a bit long, does it make sense to move it to a separate file...?

**Review by cpiemontese — APPROVED** [review]
One ignorable comment

File: .github/workflows/cd.yml
Comment (MaeIsBad) [reply .github/workflows/cd.yml:1]: I would keep it for now at least, I don't really see a good different place to put it

File: src/main.rs [hunk 1]
```diff
@@ -16,8 +17,57 @@ use localauth0::{controller, APP_NAME};
...
+        .unwrap()
+        .block_on(async {
+            let config = Config::load_or_default();
+
+            let client = reqwest::Client::builder()
+                .danger_accept_invalid_certs(true)
+                .build()
+                .unwrap();
+
+            let endpoint_http = format!("http://127.0.0.1:{}/healthcheck", config.http().port());
```
Comment (cottinisimone) [inline src/main.rs:55]: i think that `/check` might be more conforming to primait standards

File: src/main.rs [hunk 2]
```diff
@@ -16,8 +17,57 @@ use localauth0::{controller, APP_NAME};
...
+        }
+        _ => {
+            println!("{endpoint} ERROR {status:?}");
+            false
+        }
+    }
+}
+
+fn healthcheck() -> Result<(), String> {
+    tokio::runtime::Builder::new_current_thread()
```
Comment (cottinisimone) [inline src/main.rs:43]: This is another reason why i prefer having a different binary

File: src/main.rs [hunk 3]
```diff
@@ -16,8 +17,57 @@ use localauth0::{controller, APP_NAME};
 // Singleton logger. Used to free user from manually passing Logger objects around.
 static LOGGER_GUARD: GuardLoggerCell = GuardLoggerCell::new();
 
+fn main() -> Result<(), Box<dyn Error>> {
+    match std::env::args().nth(1).as_deref() {
```
[the remark this answers is not in the document]
Comment (MaeIsBad) [reply src/main.rs:21]: That was my initial implementation but that means compiling two separate versions of localauth0 which doubles the binary size

File: src/main.rs [hunk 4]
```diff
@@ -16,8 +17,57 @@ use localauth0::{controller, APP_NAME};
...
+        _ => Ok(server()?),
+    }
+}
+
+async fn is_endpoint_healthy(client: &reqwest::Client, endpoint: String) -> bool {
+    let status = client.get(&endpoint).send().await.map(|r| r.error_for_status());
+
+    match status {
+        Ok(Ok(_)) => {
+            println!("{endpoint} OK");
```
[the remark this answers is not in the document]
Comment (MaeIsBad) [reply src/main.rs:32]: No, I think it's good to leave a lot if only one of the endpoints fails to start, or something similar

File: .github/workflows/cd.yml [hunk 5]
```diff
@@ -67,20 +68,32 @@ jobs:
...
+              localauth0_binary="./localauth0-$rust_arch/localauth0"
+              chmod +x "$localauth0_binary"
+
+              ctr="$(buildah from --arch $arch $base)"
+              buildah copy "$ctr" "$localauth0_binary" "/localauth0"
+              buildah copy "$ctr" ./web/ /web/dist
+              buildah config \
+                --env 'RUST_LOG=error,localauth0=info' \
+                --cmd '["/localauth0"]' \
+                --healthcheck 'CMD /localauth0 healthcheck'
```
[the remark this answers is not in the document]
Comment (MaeIsBad) [reply .github/workflows/cd.yml:89]: I don't understand? It's useful to configure this, users can overwrite it if they want to

https://docs.docker.com/reference/dockerfile/#healthcheck

File: src/main.rs (hunk 2 above)
Comment (MaeIsBad) [reply src/main.rs:43]: Wdym?

File: src/main.rs (hunk 1 above)
Comment (MaeIsBad) [reply src/main.rs]: Done

File: src/main.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (cottinisimone) [reply src/main.rs:21]: Really? And what about a workspace with the bin with reqwest only?

File: src/main.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (cottinisimone) [reply src/main.rs:32]: Why not tracing::info or log::info?

File: src/main.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (MaeIsBad) [reply src/main.rs:21]: That's still going to include another copy of libc :/

File: src/main.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (MaeIsBad) [reply src/main.rs:32]: I wanted the healthcheck to be as minimal as possible

**Review by cpiemontese — APPROVED** [review]

**Review by cottinisimone — APPROVED** [review]

**Review by cpiemontese — APPROVED** [review]

**Pull request merged**